### PR TITLE
✨ [#065] - Assign punctuality bonus group to employee 🏆

### LINKED ISSUE #143: 👤 [#065] Assign Bonus Group to Employee
# 👤 Task #065: Assign Bonus Group to Employee

## 📖 Story

**English:**
As an Admin, I want to assign a punctuality bonus group to an employee with an effective date, so the system knows which weekly bonus amount applies when calculating their punctuality.

**Español:**
Como Admin, quiero asignar un grupo de bono de puntualidad a un empleado con fecha de vigencia, para que el sistema sepa qué monto semanal aplica al calcular su puntualidad.

---

## ✅ Backend Tasks

- [ ] 📂 Migration `create_employee_bonus_configs_table` — employee_id (FK), punctuality_bonus_group_id (FK), effective_from (date), effective_to (date nullable), timestamps
- [ ] 🔧 `EmployeeBonusConfig` model — scope `effective(date)`, `belongsTo(Employee)`, `belongsTo(PunctualityBonusGroup)`
- [ ] 🌐 `POST /api/v1/employees/{id}/bonus-config` — AssignBonusConfigController; on create auto-closes previous config
- [ ] 🌐 `GET /api/v1/employees/{id}/bonus-config` — current config + history
- [ ] 🧪 Feature tests: assign, auto-close previous, retrieve current

## ✅ Frontend Tasks

- [ ] 📝 Add `EmployeeBonusConfig` type to `src/types/attendance-payroll.ts`
- [ ] 🔧 `assignBonusConfig(employeeId, data)` + `getBonusConfig(employeeId)` in `src/services/config.service.ts`
- [ ] 📱 **Bonus config section** in Employee Detail → Configuration tab — shows current group (name, daily amount); "Cambiar grupo" button opens assignment form
- [ ] 📱 **Assignment form** — bonus group selector (populated from #064), effective_from date
- [ ] 🔧 `useEmployeeBonusConfig(employeeId)` hook

---

## 🎯 Acceptance Criteria

- [ ] Admin can assign a bonus group to an employee from their profile
- [ ] Current assignment shows the group name and computed daily amount
- [ ] Assigning a new group automatically closes the previous one

---

## 🔗 References

- **Backlog:** AP-025, AP-030 · RF-34

---

## ⏱️ Estimates

- **Optimistic:** `2h` · **Pessimistic:** `3h`

### PR DESCRIPTION
**Pull request by pakodiazdev**
## Summary

- Adds `employee_bonus_configs` table with effective-dated history (employee_id → punctuality_bonus_group_id, effective_from, effective_to)
- `GET /api/v1/employees/{id}/bonus-config` — returns full history ordered newest first
- `POST /api/v1/employees/{id}/bonus-config` — assigns a group; auto-closes the previous open config the day before
- `BonusConfigSection` component in the Employee Detail panel — shows current group (name, weekly/daily amounts) plus assignment form and history
- `useEmployeeBonusConfig` + `useAssignBonusConfig` hooks with success/error toasts

## Test plan

- [ ] 11 PHPUnit Feature tests (happy path, auth, permission, validation, auto-close logic)
- [ ] 2 PHPUnit Unit tests (scopeEffective scope behavior)
- [ ] 8 Vitest unit tests (API methods + hooks coverage)
- [ ] Cypress E2E: admin opens María García detail → assigns a bonus group → toast + active card shown
- [ ] All existing tests still pass

## Closes

Closes #143

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

<a href="https://app.devin.ai/review/pakodiazdev/sushigo/pull/144" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open in Devin Review
  </picture>
</a>

### CODE REVIEW

File: code/api/app/Http/Controllers/Api/V1/Punctuality/AssignBonusConfigController.php [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+            'punctuality_bonus_group_id' => $group->id,
+            'effective_from' => $effectiveFrom->toDateString(),
+            'effective_to' => null,
+        ]);
+
+        $config->load('bonusGroup');
+
+        return (new EmployeeBonusConfigResource($config))
+            ->response()
+            ->setStatusCode(201);
```
[the remark this answers is not in the document]
Comment (pakodiazdev) [reply code/api/app/Http/Controllers/Api/V1/Punctuality/AssignBonusConfigController.php]: Fixed in commit 4750f74. Changed to `(new EmployeeBonusConfigResource($config))->setStatusCode(201)` which calls `BaseResource::setStatusCode()` first, ensuring both the HTTP status and the JSON envelope `status` field are 201. Added a dedicated test `response_status_field_matches_http_201` that asserts `assertJsonPath('status', 201)` to prevent regression.

File: code/api/app/Http/Controllers/Api/V1/Punctuality/AssignBonusConfigController.php [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    public function __invoke(AssignBonusConfigRequest $request, Employee $employee): JsonResponse
+    {
+        $group = PunctualityBonusGroup::where('public_id', $request->bonus_group_id)->firstOrFail();
+        $effectiveFrom = Carbon::parse($request->effective_from)->startOfDay();
+
+        // Close the current open config the day before
+        $employee->bonusConfigs()
+            ->whereNull('effective_to')
+            ->where('effective_from', '<', $effectiveFrom->toDateString())
+            ->update(['effective_to' => $effectiveFrom->copy()->subDay()->toDateString()]);
```
[the remark this answers is not in the document]
Comment (pakodiazdev) [reply code/api/app/Http/Controllers/Api/V1/Punctuality/AssignBonusConfigController.php]: Fixed in commit 4750f74. Added a delete step before the close step:

```php
// Delete future open configs (effective_from >= new date) — superseded by this assignment
$employee->bonusConfigs()
    ->whereNull('effective_to')
    ->where('effective_from', '>=', $effectiveFrom->toDateString())
    ->delete();
```

This removes any open config with effective_from >= the new date (they haven't taken effect yet and are superseded). The existing close step (effective_from < new date) remains unchanged. Added test `assigning_with_earlier_date_deletes_future_open_configs` that reproduces the exact overlap scenario described.

**Comment by pakodiazdev** [comment]
## PR #144 — Follow-up Report

### Review Comments

| # | Reviewer | File / Location | Comment summary | Decision | Action taken |
|---|---|---|---|---|---|
| D1 | Devin | `AssignBonusConfigController.php:36` | `->response()->setStatusCode(201)` bypasses BaseResource envelope — JSON body keeps `"status": 200` | ✅ ATENDER | Fixed: use `->setStatusCode(201)` on resource directly. Test added: `response_status_field_matches_http_201`. Commit `4750f74` |
| D2 | Devin | `AssignBonusConfigController.php:24` | Assigning with earlier date than existing future open config creates overlapping active periods | ✅ ATENDER | Fixed: added delete step for open configs with `effective_from >= $effectiveFrom` before creating new assignment. Test added: `assigning_with_earlier_date_deletes_future_open_configs`. Commit `4750f74` |

- Total comments: 2
- Attended (fixed): 2
- Attended (already handled): 0
- Discarded: 0

### Commit Convention

| Commit SHA | Subject | ✅ Opens emoji | ✅ Issue [#NNN] | ✅ Dash ` - ` | ✅ Closes emoji | ✅ Body bullets emoji | Verdict |
|---|---|---|---|---|---|---|---|
| `987fb80` | ✨ [#065] - Assign punctuality bonus group to employee 🏆 | ✅ | ✅ | ✅ | ✅ | ✅ | ✅ PASS |
| `d63da08` | 🔧 [#065] - Move task #065 to completed 📂 | ✅ | ✅ | ✅ | ✅ | ✅ | ✅ PASS |
| `4750f74` | 🔨 [#065] - Address PR review: response envelope, overlap guard, frontend coverage 🛡️ | ✅ | ✅ | ✅ | ✅ | ✅ | ✅ PASS |

- Total commits: 3
- Compliant: 3
- Violations: 0

### SonarCloud

| Component | Quality Gate | Coverage | New Issues | Duplication |
|---|---|---|---|---|
| sushigo-api | ✅ Passed | 96.2% | 0 | 0.0% |
| sushigo-webapp | ❌ Failed (was 16.4%) → ✅ Fixed: added 16 Vitest tests for `BonusConfigSection` + `useBonusConfigSection` | — | 0 | — |

- API Quality Gate: ✅ Passed (96.2% coverage, 0 new issues)
- Webapp Quality Gate: was ❌ Failed (16.4% < 80%). Fixed with `4750f74` — 16 new Vitest tests covering `bonus-config-section.tsx` and `use-bonus-config-section.ts`. SonarCloud re-analysis pending.

### Linters

- Backend Pint: ✅ 566 files, 0 errors
- Frontend ESLint: ✅ 0 errors (50 pre-existing warnings, none in touched files)
- Frontend TypeScript: ✅ 0 errors

### Test Suite

- Existing tests (no regressions): ✅ — 174 webapp tests pass, 2738 assertions
- New API tests: 15 passing (13 feature + 2 unit scope)
- New frontend tests: 24 new Vitest tests (8 API/hooks + 8 hook unit + 8 component render)
- New API code coverage: 96.2% (≥80% ✅)
- New webapp code coverage: pending SonarCloud re-analysis (locally: 100% on new service files, component + hook fully tested)

### Overall Status

⚠️ NEEDS SONARCLOUD CONFIRMATION — all code fixes applied, tests pass, linters clean. Waiting for SonarCloud to re-analyze `4750f74` on the webapp side to confirm coverage gate passes.

### Resolved Issues

1. ~~D1: BaseResource status envelope inconsistency~~ — Fixed `4750f74`
2. ~~D2: Overlapping active periods on backdated assignment~~ — Fixed `4750f74`
3. ~~SonarCloud webapp coverage 16.4%~~ — Fixed `4750f74` (16 new tests added)

---
*Review performed by Claude Code · commit `4750f74`*

File: code/api/app/Http/Controllers/Api/V1/Punctuality/AssignBonusConfigController.php [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+        $config = $employee->bonusConfigs()->create([
+            'punctuality_bonus_group_id' => $group->id,
+            'effective_from' => $effectiveFrom->toDateString(),
+            'effective_to' => null,
+        ]);
```
[the remark this answers is not in the document]
Comment (pakodiazdev) [reply code/api/app/Http/Controllers/Api/V1/Punctuality/AssignBonusConfigController.php]: Fixed in commit acd90a6. Wrapped the three sequential DB operations (delete future configs, close previous, create new) inside `DB::transaction()` so any failure rolls back all steps atomically — consistent with the pattern used in `UpdatePunctualityRangesController`, `DeactivateEmployeeController`, and `RehireEmployeeController`. The `config` is returned from the closure and `load('bonusGroup')` runs after the transaction commits.

File: code/api/app/Http/Controllers/Api/V1/Punctuality/GetEmployeeBonusConfigController.php [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+class GetEmployeeBonusConfigController extends Controller
+{
+    public function __invoke(Employee $employee): ResourceCollection
+    {
+        $configs = $employee->bonusConfigs()
+            ->with('bonusGroup')
+            ->orderBy('effective_from', 'desc')
+            ->get();
+
+        return EmployeeBonusConfigResource::collection($configs);
```
[the remark this answers is not in the document]
Comment (pakodiazdev) [reply code/api/app/Http/Controllers/Api/V1/Punctuality/GetEmployeeBonusConfigController.php]: Fixed in commit acd90a6. Changed `GetEmployeeBonusConfigController` to resolve the collection and wrap it in `ResponseEntity`, producing the standard `{ status: 200, data: [...], meta: null }` envelope — matching the pattern in `ListPunctualityBonusGroupsController` and every other list endpoint. Removed the unused `ResourceCollection` import.

File: code/api/app/Http/Controllers/Api/V1/Punctuality/AssignBonusConfigController.php [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+            $employee->bonusConfigs()
+                ->whereNull('effective_to')
+                ->where('effective_from', '>=', $effectiveFrom->toDateString())
+                ->delete();
+
+            // Close the open config that started before the new effective date
+            $employee->bonusConfigs()
+                ->whereNull('effective_to')
+                ->where('effective_from', '<', $effectiveFrom->toDateString())
+                ->update(['effective_to' => $effectiveFrom->copy()->subDay()->toDateString()]);
```
[the remark this answers is not in the document]
Comment (pakodiazdev) [reply code/api/app/Http/Controllers/Api/V1/Punctuality/AssignBonusConfigController.php:31]: Fixed in commit be9668a. Added `lockForUpdate()` to the close step inside the transaction, consistent with the pattern used in `DeactivateEmployeeController` and `RehireEmployeeController`. The delete step (for future configs) runs first and doesn't need a lock since it's unconditional; only the update that closes the previous open record needs the row lock.

File: code/api/app/Http/Controllers/Api/V1/Punctuality/AssignBonusConfigController.php [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+            // Delete future open configs (effective_from >= new date) — superseded by this assignment
+            $employee->bonusConfigs()
+                ->whereNull('effective_to')
+                ->where('effective_from', '>=', $effectiveFrom->toDateString())
+                ->delete();
+
+            // Close the open config that started before the new effective date
+            $employee->bonusConfigs()
+                ->whereNull('effective_to')
+                ->where('effective_from', '<', $effectiveFrom->toDateString())
```
[the remark this answers is not in the document]
Comment (pakodiazdev) [reply code/api/app/Http/Controllers/Api/V1/Punctuality/AssignBonusConfigController.php]: Fixed in commit be9668a. Removed the `whereNull('effective_to')` filter from the delete step — now it deletes ALL configs with `effective_from >= $effectiveFrom`, regardless of whether they are open or already closed by a later assignment. Added test `assigning_with_earlier_date_also_deletes_closed_future_configs` that reproduces the exact Jan→Apr, Apr→May, post-Mar scenario described.

File: code/api/app/Http/Requests/Punctuality/AssignBonusConfigRequest.php [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+    public function authorize(): bool
+    {
+        return true;
+    }
+
+    public function rules(): array
+    {
+        return [
+            'bonus_group_id' => ['required', 'string', 'exists:punctuality_bonus_groups,public_id'],
+            'effective_from' => ['required', 'date'],
```
[the remark this answers is not in the document]
Comment (pakodiazdev) [reply code/api/app/Http/Requests/Punctuality/AssignBonusConfigRequest.php]: Fixed in commit be9668a. Replaced the bare `exists` rule with `Rule::exists('punctuality_bonus_groups', 'public_id')->where('is_active', true)`, so assigning a retired group returns 422 even when calling the API directly. Added test `assign_bonus_config_rejects_inactive_bonus_group` to prevent regression.

File: code/api/routes/api.php [hunk 8]
```diff
@@ -337,6 +344,8 @@
     Route::middleware('auth:api')->prefix('punctuality')->name('punctuality.')->group(function () {
         Route::get('/ranges', ListPunctualityRangesController::class)->name('ranges.index')->middleware('permission:punctuality.manage');
         Route::put('/ranges', UpdatePunctualityRangesController::class)->name('ranges.update')->middleware('permission:punctuality.manage');
+        Route::get('/bonus-groups', ListPunctualityBonusGroupsController::class)->name('bonus-groups.index')->middleware('permission:punctuality.manage');
```
[the remark this answers is not in the document]
Comment (pakodiazdev) [reply code/api/routes/api.php]: Fixed in commit be9668a. Changed the middleware on `GET /punctuality/bonus-groups` to `permission:punctuality.manage|employees.update` so users who can assign bonus configs (employees.update) can also load the group list needed by the assignment form. Added test `user_with_employees_update_can_list_bonus_groups` to verify the new access path.

File: code/webapp/src/components/employees/use-bonus-config-section.ts [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+  const { data: configs, isLoading: isLoadingConfigs } = useEmployeeBonusConfig(employeeId)
+  const { data: groups, isLoading: isLoadingGroups } = useBonusGroups()
+  const assign = useAssignBonusConfig(employeeId)
+
+  const form = useForm<AssignFormValues>({
+    resolver: zodResolver(assignSchema),
+    defaultValues: { bonus_group_id: '', effective_from: '' },
+  })
+
+  const current = configs?.[0] ?? null
```
[the remark this answers is not in the document]
Comment (pakodiazdev) [reply code/webapp/src/components/employees/use-bonus-config-section.ts]: Fixed in commit be9668a. Changed `current` derivation from `configs?.[0]` to a `.find()` that matches configs where `effective_from <= today && (effective_to === null || effective_to >= today)`. A future scheduled config no longer shows as Activo. Added two Vitest tests covering the future-only and all-expired edge cases.

File: code/api/app/Http/Requests/Punctuality/CreatePunctualityBonusGroupRequest.php [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+ *     schema="CreatePunctualityBonusGroupRequest",
+ *     required={"name","weekly_bonus_amount","working_days_divisor"},
+ *
+ *     @OA\Property(property="name", type="string", maxLength=50, example="Grupo $110 (÷6)"),
+ *     @OA\Property(property="weekly_bonus_amount", type="number", format="float", minimum=0.01, example=110.00),
+ *     @OA\Property(property="working_days_divisor", type="integer", minimum=1, example=6)
+ * )
+ */
+class CreatePunctualityBonusGroupRequest extends FormRequest
+{
```
[the remark this answers is not in the document]
Comment (pakodiazdev) [reply code/api/app/Http/Requests/Punctuality/CreatePunctualityBonusGroupRequest.php:18]: Fixed in commit be9668a. Added `public function authorize(): bool { return true; }` to `CreatePunctualityBonusGroupRequest`. Without it, Laravel's default returns `false` and denies every request with 403 before the controller runs. The existing `admin_can_create_bonus_group` test now validates this works end-to-end.
